Authn docs (#1151)
emqtt/emq-docs-cn: en_US/security/authn/http.md

# HTTP

HTTP authenticator delegates authentication to a custom HTTP API.

## Authentication principle

> * In authenticator settings, an HTTP request pattern is specified.

Reviewer: ```suggestion
* In authenticator settings, a HTTP request pattern is specified.
```

Reviewer 2: `an` is correct in this case

* When an MQTT client connects to the broker, the configured request template is rendered and the resulting request is emitted.

> * Received 200 or 204 HTTP status is interpreted as authentication succes. Other statuses indicate authentication failure.

Reviewer: ```suggestion
* Received 200 or 204 HTTP status is interpreted as authentication success. Other statuses indicate authentication failure.
```

> * Received 200 or 204 HTTP status is interpreted as authentication succes. Other statuses indicate authentication failure.

Reviewer 2: ```suggestion
* Receiving a 200 or 204 HTTP status is interpreted as authentication success. Other statuses indicate authentication failure.
```

> Successful HTTP status can also contain a JSON or www-form-urlencoded map with `is_superuser` boolean field

Reviewer 2: ```suggestion
A successful HTTP response can also contain a JSON or www-form-urlencoded map with `is_superuser` boolean field
```

> that indicates superuser priveleges for the client.

Reviewer: ```suggestion
that indicates superuser privileges for the client.
```

For untrusted environments HTTPS should be used.
:::

## Configuration

HTTP authentication is identified with `mechanism = http` and `backend = built_in_database`.

HTTP `POST` and `GET` requests are supported. Each of them has some specific options.

Example of an HTTP authenticator configured with `POST` request:

```hocon
{
    mechanism = password_based
    backend = http
    enable = true

    method = post
    url = "http://127.0.0.1:32333/auth/${peercert}?clientid=${clientid}"
    body {
        username = "${username}"
        password = "${password}"
    }
    headers {
        "Content-Type" = "application/json"
        "X-Request-Source" = "EMQX"
    }
}
```

Example of an HTTP authenticator configured with `GET` request:

```hocon
{
    mechanism = password_based
    backend = http
    enable = true

    method = get
    url = "http://127.0.0.1:32333/auth"
    body {
        username = "${username}"
        password = "${password}"
    }
    headers {
        "X-Request-Source" = "EMQX"
    }
}
```

### `method`

Required field with possible values `get` or `post`. Denoting the corresponding HTTP request method used.

### `url`

> HTTP url for external authentication requests, required. It may cantain [placeholders](./authn.md#authentication-placeholders).

Reviewer: ```suggestion
HTTP url for external authentication requests, required. It may contain [placeholders](./authn.md#authentication-placeholders).
```
